pandas/core/internals/blocks.py

```diff
@@ -596,6 +598,21 @@ def astype(self, dtype, copy: bool = False, errors: str = "raise"):
 
             return self.make_block(Categorical(self.values, dtype=dtype))
 
+        elif (  # GH8628
+            is_categorical_dtype(self.values.dtype)
+            and not (
+                is_object_dtype(dtype)
+                or is_string_like_dtype(dtype)
+                or is_extension_array_dtype(dtype)
+                or is_datetime_or_timedelta_dtype(dtype)
+            )
+            and copy is True
+        ):
```

Comment:
this seems really convoluted, in a method that is already too complicated as it is (xref #22369)

Do you have a good idea where the perf improvement comes from?  e.g. could we push this down into `Categorical.astype`?

Reply:
Agreed that the amount of special casing here is not good (and even with that a bunch of tests are still failing)

The perf improvement is from astyping just the category labels instead of astyping each array entry separately.

`Categorical.astype` seems like a good location for this - will give that a go